src/mlpack/bindings/cli/add_to_po.hpp

```diff
@@ -14,95 +13,85 @@
 #define MLPACK_BINDINGS_CLI_ADD_TO_PO_HPP
 
 #include <mlpack/core/util/param_data.hpp>
-#include <boost/program_options.hpp>
 #include <mlpack/core/util/is_std_vector.hpp>
 #include "map_parameter_name.hpp"
 
+#include "third_party/CLI/CLI11.hpp"
+
 namespace mlpack {
 namespace bindings {
 namespace cli {
 
 /**
- * Add a non-vector option to boost::program_options.
- *
- * @param boostName The name of the option to add to boost::program_options.
- * @param descr Description string for parameter.
- * @param desc Options description to add parameter to.
- */
-template<typename T>
-void AddToPO(const std::string& boostName,
-             const std::string& descr,
-             boost::program_options::options_description& desc,
-             const typename boost::disable_if<util::IsStdVector<T>>::type* = 0,
-             const typename boost::disable_if<std::is_same<T, bool>>::type* = 0)
-{
-  desc.add_options()(boostName.c_str(), boost::program_options::value<T>(),
-      descr.c_str());
-}
-
-/**
- * Add a vector option to boost::program_options.  This overload will use the
- * multitoken() option.
- *
- * @param boostName The name of the option to add to boost::program_options.
+ * Add an option to CLI11.
+ * 
+ * @param cliName The name of the option to add to CLI11.
  * @param descr Description string for parameter.
  * @param desc Options description to add parameter to.
  */
 template<typename T>
-void AddToPO(const std::string& boostName,
-             const std::string& descr,
-             boost::program_options::options_description& desc,
-             const typename boost::enable_if<util::IsStdVector<T>>::type* = 0,
+void AddToPO(const std::string& cliName,
+             util::ParamData& param,
+             CLI::App& app,
              const typename boost::disable_if<std::is_same<T, bool>>::type* = 0)
 {
-  desc.add_options()(boostName.c_str(),
-      boost::program_options::value<T>()->multitoken(), descr.c_str());
+  app.add_option_function<T>(cliName.c_str(),
+      [&param](const T& value)
+      {
+        param.value = value;
+        param.wasPassed = true;
```

Comment:
I think here is our issue (somehow).  I took a look at the backtrace, and it seems like `data::Load()` is being called with a filename that is (somehow) an invalid string.  `data::Load()` is going to be called, in this context, from inside `GetParam<N>()`.  This code seems to work for when `N` is `arma::mat`, but the failing case we have is when `N` is actually a categorical matrix (i.e. `N = std::tuple<mlpack::data::DatasetInfo, arma::mat>`).

I think the reason is that, in this particular code, `T` in this function is `std::string` (because `ParameterType<typename std::remove_pointer<std::tuple<mlpack::data::DatasetInfo, arma::mat>>>` is `std::string`).  So that means we are assigning `value` (of type `std::string`) to `param.value`... but later in `GetParam<>`, we are expecting that `param.value` holds a `std::tuple<std::tuple<mlpack::data::DatasetInfo, arma::mat>, std::string>` (i.e. see line 84 of `get_param.hpp`.

So maybe I was wrong earlier and what we actually have to do is have different overloads for `AddToPO()` when `T` in the bottom-most overload of `AddToPO()` is `arma::mat` or `std::tuple<mlpack::data::DatasetInfo, arma::mat>`, and in those cases, we can call `GetRawParam<T>(d) = value`.

This is pretty complex infrastructure so let me know if there's anything else I can clarify about what I wrote.  It might take a little while of reading through the code to understand everything I wrote here. :+1: